lib/spack/spack/patch.py

```diff
@@ -150,50 +193,90 @@ def __init__(self, pkg, relative_path, level, working_dir, ordering_key=None):
 
         super().__init__(pkg, abs_path, level, working_dir)
         self.path = abs_path
-        self._sha256 = None
+        self._sha256 = sha256
         self.ordering_key = ordering_key
 
     @property
-    def sha256(self):
-        if self._sha256 is None:
+    def sha256(self) -> Optional[str]:
+        """Get the patch checksum.
+
+        Returns:
+            The sha256 of the patch file.
+        """
+        if self._sha256 is None and self.path is not None:
             self._sha256 = checksum(hashlib.sha256, self.path)
         return self._sha256
 
-    def to_dict(self):
-        return llnl.util.lang.union_dicts(super().to_dict(), {"relative_path": self.relative_path})
+    @sha256.setter
+    def sha256(self, value: Optional[str]) -> None:
+        """Set the patch checksum.
+
+        Args:
+            value: the sha256
+        """
+        self._sha256 = value
+
+    def to_dict(self) -> Dict[str, Any]:
+        """Dictionary representation of the patch.
+
+        Returns:
+            A dictionary representation.
+        """
+        data = super().to_dict()
+        data["relative_path"] = self.relative_path
+        return data
 
 
 class UrlPatch(Patch):
-    """Describes a patch that is retrieved from a URL.
-
-    Arguments:
-        pkg (str): the package that owns the patch
-        url (str): URL where the patch can be fetched
-        level (int): level to pass to patch command
-        working_dir (str): path within the source directory where patch
-            should be applied
-    """
+    """Describes a patch that is retrieved from a URL."""
+
+    def __init__(
+        self,
+        pkg: "spack.package_base.PackageBase",
+        url: str,
+        level: int = 1,
+        working_dir: str = ".",
+        ordering_key: Optional[Tuple[str, int]] = None,
+        sha256: Optional[str] = None,
```

Comment:
Same concern as before on this argument

Reply:
For `UrlPatch`, isn't `sha256` required? This should actually be `str`, not `Optional[str]` imo.

Reply:
Yeah we do require it, so `str` here IMO

Reply:
It's hard to use str here because the `patch` directive is dynamic.

Reply:
see what you think of [ce01485](https://github.com/spack/spack/pull/42811/commits/ce014859f780a5ab85627436fbd1778a224ae805)... I'm not sure if I like it or not.